yara-x/src/modules/macho.rs

```diff
@@ -0,0 +1,1801 @@
+
+    // Print Segment Commands
+    for segment in &macho_proto.segments {
+        println!("Segment Commands:");
+        println!("Command: {}", print_option_hex(segment.cmd));
+        println!("Command Size: {}", print_option(segment.cmdsize));
+        println!("Segment Name: {}", print_option(segment.segname.as_ref()));
+        println!("VM Address: {}", print_option_hex(segment.vmaddr));
+        println!("VM Size: {}", print_option_hex(segment.vmsize));
+        println!("File Offset: {}", print_option(segment.fileoff));
+        println!("File Size: {}", print_option(segment.filesize));
+        println!("Max Protection: {}", print_option_hex(segment.maxprot));
+        println!("Init Protection: {}", print_option_hex(segment.initprot));
+        println!("Number of Sections: {}", print_option(segment.nsects));
+        println!("Flags: {}", print_option_hex(segment.flags));
+        // Print nested Segment Sections
+        for section in &segment.sections {
+            println!("Sections:");
+            println!(
+                "Segment Name: {}",
+                print_option(section.segname.as_ref())
+            );
+            println!(
+                "Section Name: {}",
+                print_option(section.sectname.as_ref())
+            );
+            println!("Address: {}", print_option_hex(section.addr));
+            println!("Size: {}", print_option_hex(section.size));
+            println!("Offset: {}", print_option(section.offset));
+            println!("Alignment: {}", print_option(section.align));
+            println!("Relocation Offset: {}", print_option(section.reloff));
+            println!(
+                "Number of Relocations: {}",
+                print_option(section.nreloc)
+            );
+            println!("Flags: {}", print_option_hex(section.flags));
+            println!("Reserved 1: {}", print_option(section.reserved1));
+            println!("Reserved 2: {}", print_option(section.reserved2));
+            println!("Reserved 3: {}", print_option(section.reserved3));
+            println!();
+        }
+        println!();
+    }
+
+    // Print Number of Segments
+    println!(
+        "Number of segments: {}",
+        print_option(macho_proto.number_of_segments)
+    );
+
+    // Print Entry Point
+    println!("Entry Point: {}", print_option(macho_proto.entry_point));
+
+    // Print Stack Size
+    println!("Stack Size: {}", print_option(macho_proto.stack_size));
+}
+
+#[module_main]
+fn main(ctx: &ScanContext) -> Macho {
+    // Create an empty instance of the Mach-O protobuf
+    let mut macho_proto = Macho::new();
+
+    // Set constants before parsing the binary
+    set_definitions(&mut macho_proto);
+
+    // Get a &[u8] slice with the content of the file being scanned.
+    let data = ctx.scanned_data();
+
+    // If data is too short to be valid Mach-O file, return empty protobuf
+    if data.len() < VALID_MACHO_LENGTH {
+        eprintln!("{}", MachoError::FileTooSmall);
```

Comment:
Thank you @plusvic also for this. I played with a [log](https://docs.rs/log/latest/log/) crate while developing this module but got a bit lost in other logs while using it for printing `protobuf` [fields](https://github.com/TommYDeeee/yara-x/blob/108e048d1f431f6d7b986ad7d17f26fcb08fb0d7/yara-x/src/modules/macho.rs#L1841) to check if everything is ok so far. Also this will be removed once I finish parsing and start with tests for this module. 

In terms of error handling like you mentioned I am not sure what is expected. Should each module print out their errors using `error` categories, should this be somehow handled for modules all togather? Or should I somehow propagate the error upwards until it is processed by `yara-x-cli` [handler](https://github.com/VirusTotal/yara-x/blob/2745dfacc757648c9fc0a27320d148646ebe79ee/yara-x-cli/src/main.rs#L77) like rest of the code. I found the latest one to be the best and wanted to ask if it can be somehow propagated there. So far I didn't find a way how to do it as `module_main` macro is expecting just protobuf as a result.

Thank you for any insight into this problem.

Reply:
I think there are two different discussions going on here: one is about error handling, and the other about logging. 

With respect to error handling, the module's main function always return a protobuf alone, so we can say that the module's main function must be infallible. If something goes wrong while parsing the data, the protobuf is simply empty, or half-populated. The module's main function could return an error as well, but that would be of little use, as YARA itself won't do anything with that error. YARA won't abort the scanning or report that error to the user in any form, that's why the main function only returns the protobuf, in order to reinforce the message that the module is not expected to fail, and if it does, it must fail silently.

In the other hand we have logging, which can be useful for inspecting the module's behavior. You can use logging for providing information about what's happening with the module, but that information is only for debugging purposes, whether to use `debug`, `info`, `warn` or `error` categories depends on what are you logging exactly, but my recommendation is trying to log only the most important information as `info`, `warn` and `error`. And use `debug` for more verbose stuff.